Add session recording summarizer documentation (#58572)
gravitational/teleport: docs/pages/identity-security/session-summaries.mdx

is granted by default through the built-in `editor` role.

You can either use the built-in roles or create a new one to make sure you have access to all necessary resources. To create a new role:

1. Create a `summarizer-admin.yaml` file with the following content:

   ```yaml
   kind: role
   metadata:
     name: summarizer-admin
   spec:
     allow:
       rules:
       - resources:
         - inference_model
         - inference_secret
         - inference_policy
         verbs:
         - read
         - list
         - create
         - update
         - delete
       - resources:
         - session
         verbs:
         - read
         - list
   version: v7
   ```

1. Apply your changes:

   ```code
   $ tctl create -f summarizer-admin.yaml
   ```

   (!docs/pages/includes/create-role-using-web.mdx!)

1. (!docs/pages/includes/add-role-to-user.mdx role="summarizer-admin"!)

## Step 2: Create the inference configuration resources

Let's start by defining an `inference_model` resource. It tells Teleport how to
summarize a session by pointing it to a language model API. In particular, it
contains an API-specific model name. Note that the *Teleport inference model

> name* and *API-specific model name* are two different things. This allows, for

Author: @tigrato I'm not sure what your intention is here. Did you want to deemphasize both terms? Emphasize the word "and"? The reason why I applied this formatting is to emphasize the difference between these two.

> name* and *API-specific model name* are two different things. This allows, for

Author: It's italics, not bold. @ptgott, do we have any clear guidelines on using italics or bold for emphasis in such cases?

> name* and *API-specific model name* are two different things. This allows, for

Reviewer: We tend to use bold to introduce new keywords. I think italics are fine here, since this is a pretty conventional use case for them, but I would just check the preview site for whether the single word between the two italicized phrases is enough to visually separate them.

example, using the same OpenAI model with different temperature settings or
using different API gateways by creating multiple Teleport inference models
with the same OpenAI model name.

To communicate with OpenAI, an inference model needs an API key. It's stored
separately in an `inference_secret` resource and referenced by the
`inference_model`.

<Admonition type="warning">
To protect your API key, once written, an `inference_secret` value cannot be
retrieved through `tctl` or Teleport API.
</Admonition>

Finally, we add an `inference_policy` that routes sessions to inference models
using session kinds and optional custom filters. Here are some example
scenarios that can be expressed using inference policies:

> - Only summarize sessions to your production resources using resource labels.

Reviewer 2: Yeah that would be good to include a few common examples.

appear up to a minute after the session finishes.
</Admonition>

## Limitations

Currently, Teleport doesn't attempt to summarize sessions that are larger than
the model's context window. If after tokenization, the session turns out to be
larger than that, OpenAI will simply reject the request; the input tokens used
will be wasted. To prevent this, Teleport allows you to customize the maximum
size of session that may be sent to the inference provider by using the
`spec.max_session_length_bytes` field of the `inference_model` resource. Note
that there's no way to provide an exact number of bytes for a given model here;
in our experiments, we observed between 2 and 4 bytes per token, depending on
the model and input.

## Selecting sessions to summarize

Indiscriminately summarizing all sessions in your cluster can quickly drain
your wallet, so it's possible to be specific about which sessions will be
summarized and which model will be used for inference. One of these mechanisms
— selecting sessions by kind — was mentioned already in the above example. In
addition to matching by kind, it's also possible to add a custom filter to each

> inference policy.

Author: This is covered in the next three paragraphs. What other specific examples would you like to see?

> inference policy.

Author: (I'm also going to change "it's possible to be specific" to "is recommended to be specific".)

Filters are expressed using the
[Teleport predicate language](../reference/predicate-language.mdx#matching-session-summary-inference-policies).
If an inference policy contains a filter, it will match only sessions that
match one of the indicated kinds *and* the specified filter. A typical case
would be to pick only production resources for summarizing. The following
example policy only matches Kubernetes sessions in any of the production
clusters, picked by matching cluster labels:

```yaml
kind: inference_policy
version: v1
metadata:
  name: prod-k8s-sessions
  description: Summarize production Kubernetes sessions
spec:
  kinds:
  - k8s

>   filter: 'equals(resource.metadata.labels["env"], "prod")'

Author: I checked it, but it doesn't validate. Whether it's a bug in an overzealous validation code or something else, let's reflect the status quo for now, and then we'll see.

>   filter: 'equals(resource.metadata.labels["env"], "prod")'

Reviewer 3: you are correct. I though we embedded the metadata inside resource but that's not the case

>   model: ssh-summary-model

Reviewer 3: then model is called ssh but it's filtering k8s

---
title: Session Recording Summaries

> description: Describes how to generate and view session recording summaries using a language model.

Reviewer 2: This description makes it sound like we're going to walk the user through a set of steps to generate summaries, but that mostly happens automatically.

What if we said this instead?

```suggestion
description: Describes how to use Teleport Identity security to summarize session recordings with a language model.
```

labels:
  - conceptual

>   - identity-security

Reviewer 2: ```suggestion
  - identity-security
  - AI 
```

participants, resource labels, user traits, etc.

If a match is found, Teleport proceeds to summarize the session recording. It
uses an inference model configuration indicated by the policy to generate a
summary using an external inference provider, such as OpenAI. There is no way
to run this process on demand.

There are three fundamental concepts to that underpin session summarization:

- **Inference providers** do the heavy lifting of summarization. Currently,
  Teleport only supports OpenAI or compatible inference providers.
- **Inference models** indicate *how* sessions get summarized: they are
  Teleport configuration resources that provide model settings.
- **Inference policies** indicate *what* gets summarized and which inference
  model is used in given case.

## Prerequisites

- A running Teleport Enterprise cluster v18.2.0 or later with Identity Security
  enabled.
- Access to OpenAI or OpenAI-compatible inference API, either public or
  self-hosted.

## Step 1/4: Configure access to inference configuration

In order to manage the summary inference configuration, you need a write access
to `inference_model`, `inference_secret`, and `inference_policy` resources,

> which is granted by default through the built-in `editor` role.

Reviewer 2: ```suggestion
which is granted by default through the preset `editor` role.
```

> You can either use the built-in roles or create a new one to make sure you have
> access to all necessary resources. To create a new role:

Reviewer 2: ```suggestion
Alternatively, you can create a new role to make sure you have
access to all necessary resources. To create a new role:
```

         - list
         - create
         - update
         - delete
   version: v7
   ```

1. Apply your changes:

   ```code
   $ tctl create -f summarizer-admin.yaml
   ```

   (!docs/pages/includes/create-role-using-web.mdx!)

1. (!docs/pages/includes/add-role-to-user.mdx role="summarizer-admin"!)

## Step 2/4: Create the inference configuration resources

Let's start by defining an `inference_model` resource. It tells Teleport how to
summarize a session by pointing it to a large language model (LLM) API. In
particular, it contains an API-specific model name. Note that the *Teleport
inference model name* and *API-specific model name* are two different things.
This allows, for example, using the same OpenAI model with different
temperature settings or using different API gateways by creating multiple
Teleport inference models with the same OpenAI model name.

To communicate with OpenAI, an inference model needs an API key. It's stored
separately in an `inference_secret` resource and referenced by the
`inference_model`'s `spec.openai.api_key_secret_ref` field.

<Admonition type="warning">
To protect your API key, once written, an `inference_secret` value cannot be

> retrieved through `tctl` or Teleport API. It can only be updated or deleted.

Reviewer 2: Not even with `tctl get --with-secrets`?

What happens if you try to `tctl edit` the resource?

- Use different models for summarizing shell and database sessions.

To create the required configuration resources:

1. Create an `inference-config.yaml` file with the following content:

   ```yaml
   kind: inference_model
   version: v1
   metadata:
     name: shell-summary-model
   spec:
     # "openai" is currently the only supported inference provider.
     openai:
       # In this scenario, where we connect to OpenAI directly, openai_model_id
       # needs to be a valid OpenAI model name.
       openai_model_id: gpt-4o
       # This value refers to the `inference_secret` resource name.
       api_key_secret_ref: openai-key
     # It's a good idea to limit the session length to prevent incurring cost for
     # sessions that were too long anyway to fit in the model's context window.
     # If unset, defaults to 200kB. See the Limitations section for more details.
     max_session_length_bytes: 190000

   ---

   kind: inference_secret
   version: v1
   metadata:
     name: openai-key
   spec:
     value: '<paste-your-openai-api-key-here>'

   ---

   kind: inference_policy
   version: v1
   metadata:

>      name: shell-sessions

Reviewer 2: I wonder if we should establish a best practice of naming these something like:

```suggestion
     name: 0001-openai-gpt4o
```

(Just to help reinforce the concept that you have to be thoughtful about the name because the first alphabetic match wins)

Author: I think we should rather introduce a priority field in the next version. Or perhaps switch to allowing multiple policies in a single resource.

     description: Summarize all sessions using OpenAI
   spec:
     # Session kinds affected by this policy. Allowed values: "ssh", "k8s", "db".
     kinds:
     - ssh
     - k8s
     - db
     # Name of the `inference_model` resource that will be used for summarizing
     # these sessions.

>      model: shell-summary-model

Reviewer 2: These resources all seem to reference each other by name.

Just curious if we do any sort of checking to verify that the referenced resource exists and to help catch typos.

Author: No, not in this version. :) But I keep that in mind.

   ```

1. Create the configuration resources in the backend:

   ```code
   $ tctl create -f inference-config.yaml
   ```

For full description of the inference configuration resources, see appropriate
sections of [Teleport Resources Reference](../reference/resources.mdx):

- [`inference_model`](../reference/resources.mdx#inference-model)
- [`inference_secret`](../reference/resources.mdx#inference-secret)
- [`inference_policy`](../reference/resources.mdx#inference-policy)

## Step 3/4: Configure access to session summaries.

Every user who has access to
[given session recording](../reference/access-controls/roles.mdx#rbac-for-session-recordings)
also has access to its summary through a `session` resource kind. The built-in
`access` role allows users to only view their sessions; in order to allow a

> user to view all recordings and summaries, it's easiest to use the built-in

Reviewer 2: ```suggestion
user to view all recordings and summaries, it's easiest to use the preset
```

`auditor` role.

You can either use the built-in roles or create a new one to make sure you have
access to all recordings. To create a new role:

1. Create a `summary-auditor.yaml` file with the following content:

   ```yaml
   kind: role
   metadata:
     name: summary-auditor
   spec:
     allow:
       rules:
       - resources:
         - session
         verbs:
         - read
         - list
   version: v7
   ```

1. Apply your changes:

   ```code
   $ tctl create -f summary-auditor
   ```

1. (!docs/pages/includes/add-role-to-user.mdx role="summary-auditor"!)

## Step 4/4: Conduct a session and view its summary

That's it! We are ready to conduct a test session. Connect to any of your

> cluster resources using `tsh ssh`, `kubectl`, `tsh db connect`, or a web

Reviewer 2: ```suggestion
cluster resources using `tsh ssh`, `kubectl exec`, `tsh db connect`, or a web
```

Would `tsh db connect` actually trigger a summary? My understanding is that this will only work for Postgres (and soon to be MySQL) sessions run from the Web UI.

terminal. Start an interactive session and execute any command, quit the session, and go to **Audit -> Session
Recordings**. You should see the corresponding recording tile with a summary
button ![summary button](../../img/identity-security/summary-button.png).
Clicking the summary button will show the summary.

![Session recording list screen with a summary visible](../../img/identity-security/rec-list-with-summary.png)

<Admonition type="info">
It can take a while to generate a summary, and the time depends on multiple
factors, including session size, model, and its temperature. Usually it should

> appear up to a minute after the session is uploaded to Auth service.

Reviewer 2: ```suggestion
appear up to a minute after the session is uploaded to the Auth service.
```
